### ISSUE TITLE
Why do I have to retrain the model every time?

### ISSUE THREAD

**Issue by BasilBeirouti** [DistrictDataLabs/yellowbrick#656]
**Describe the solution you'd like**
I want to be able to use ResidualsPlot and PredictionError classes without having to retrain the model every time. I want to be able to train the model on my own and pass in the y_actual, y_predicted, and X_test and make the plots. I have tried to use the draw() method but I get errors. 

**Is your feature request related to a problem? Please describe.**
Very frustrating to have to repeatedly train the model. I am having trouble even believing that this how it was designed, I hope I'm just missing something

**Examples**
I want to be able to do this:
```
X_train, X_test, y_train, y_test = train_test_split(X, y)
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_test)
resid = y_test - y_pred
resplot = ResidualsPlot()  
resplot.draw(y_test, y_pred)
errplot = PredictionError()
errplot.draw(y_test, y_pred)
```

**Comment by BasilBeirouti** [comment]
The reason I have a big dataset, it takes some time to train and I want to avoid having to train multiple times. Also, because I have so much data, the plots are just completely solid because of all the overlapping markers. So I feel it should also be possible to modify size and type of markers in both of those classes, and to choose to only plot the test data, for example.

**Comment by bbengfort** [comment]
@BasilBeirouti thank you for using Yellowbrick and for your question - it turns out that it should be fairly straight-forward to use these visualizers without retraining the model. For prediction error, try:

```python
errplot = PredictionError(lr) 
errplot.score(X_test, y_test) 
errplot.poof() 
```

This should be sufficient for drawing a complete prediction error plot since only the test data is drawn. I'm assuming the error you got when using `draw` directly was that there was no attribute `score_`, which is computed in the score method. 

The residuals plot is a bit tricker, since fit is required to plot the training data. If you are only interested in the residuals of the test data, you can use score as above. However, there is a hack you can use to get both train and test data without calling fit:

```python
resplot = ResidualsPlot(lr) 
resplot.score(X_train, y_train, train=True) 
resplot.score(X_test, y_test) 
resplot.poof() 
```

The `train argument` is [documented](http://www.scikit-yb.org/en/latest/api/regressor/residuals.html#yellowbrick.regressor.residuals.ResidualsPlot) but not widely advertised, but this should get you where you'd like to go. 

For the last part of your question; unfortunately, we haven't generalized these plots enough to allow you to modify the size and marker of the scatter plot, just the color, and opacity. If this is a feature you think is needed in these plots, please open a feature request for it and we'll get it on the development roadmap! 

You do have a few options to deal with this, however:

1. Use the `resplot.ax` to get access to the scatter plot directly and modify it (though I'm not 100% sure how to do this). 
2. Modify `matplotlib.rcParams` e.g. `rcParams["lines.markersize"]` and `rcParams["lines.marker"]`
3. Modify the opacity of the markers using the `alpha` argument, e.g. `ResidualsPlot(lr, alpha=0.45)`
4. Modify the opacity by specifying the color, e.g. `ResidualsPlot(lr, train_color='#95BB6CCC', test_color='#036798FF')`, by defining colors as RGBA values. 

I hope this helps answer your question, let me know if you're still having trouble.

**Comment by pedropalb** [comment]
Hello!

Do you plan to decouple the model object from the model evaluation functions? I mean an interface like `function(y_true, y_pred)` so one don’t have to delegate the interactions with the model to the Yellowbrick package. I believe it would generalize the package’s interface to evaluate models coming from all sorts of environments (e.g. R). This way there is no need for wrappers for Keras models, for example. The caller would be responsible to interact with the model to generate y_pred. 

I’m working with a deep neural net in tf keras for a regression task and I already have the pair (y_true, y_pred). Without such an interface I have to create a wrapper so Yellowbrick can call the predict to generate the y_pred to compare with y_true. But I already have the pair and shouldn’t need to delegate the predict to the Yellowbrick.

Thank you!

**Comment by bbengfort** [comment]
Hi @pedropalb - thanks for using Yellowbrick! We currently don't have plans to create such an interface. The issue is that Yellowbrick often takes advantage of the properties of models, particularly the learned attributes such as `coef_` in order to produce its visualizations. There are a few visualizers that can produce results without the estimator, but we've erred on the side of integrating into the ML workflow. (Note that we're still looking to produce Keras estimators for Yellowbrick - #1106). 

However, it is hopefully pretty simple to produce an "estimator" that can do what you'd like. I thought we actually had this already in the contrib module, but for some reason I can't find it - I'll investigate and get back to you. But the basic sketch is as follows:

```python
from sklearn.base import BaseEstimator
from sklearn.metrics import f1_score

class PassthroughEstimator(BaseEstimator):

    def __init__(self, y_pred, estimator_type="classifier"):
        self._y_pred = y_pred
        self._estimator_type = estimator_type        

    def fit(self, X, y=None):
        return self
    
    def predict(self, X):
        return self._y_pred
    
    def score(self, X, y=None):
        return f1_score(y, self._y_pred)
```

You may get some issues if Yellowbrick looks for an learned attribute you can't find; but you can add them to the passthrough from your Keras model to make it work. Hope that helps!

**Comment by bbengfort** [comment]
@pedropalb I think there is a blog post on this topic, but until I find the link - I wanted to mention that I've opened a PR that adds this functionality to the yellowbrick contrib module: #1189
